### PR TITLE
added a utility to clean a fixture's field name

### PR DESCRIPTION
**Pull request by biyeun**
so that bad xml is not sent down from a phone restore

addresses part of this ticket http://manage.dimagi.com/default.asp?176616#982925

did part 2 of this ticket:
- validation on the server and client side.
  examples:
  screen shot 2015-07-24 at 12 48 26 pm
  screen shot 2015-07-24 at 12 48 18 pm
  screen shot 2015-07-24 at 11 32 20 am
  screen shot 2015-07-24 at 11 32 08 am
  screen shot 2015-07-24 at 11 29 55 am

@esoergel @orangejenny cc @kaapstorm

### CODE REVIEW

File: corehq/apps/fixtures/utils.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import re
+
+
+def clean_fixture_field_name(field_name):
```
Comment (czue) [inline corehq/apps/fixtures/utils.py]: this seems like a great candidate for a few `SimpleTestCase`s

File: corehq/apps/fixtures/utils.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    """Effectively slugifies a fixture's field name so that we don't send
+    bad XML back from the phone. Ideally, the fixture name should be
+    verified as a good slug before using it.
+    """
+    pattern = '([/\\<> ])'
+    return re.sub(pattern, '_', field_name)
```
Comment (esoergel) [inline corehq/apps/fixtures/utils.py]: I think these field names need to also be prevented from starting with `"xml"`.  That's against the xml spec, although I don't know whether or not that'll actually fail hard in CommCare.

File: corehq/apps/fixtures/tests.py [hunk 3]
```diff
@@ -165,3 +166,142 @@ def test_get_fixture_data_types_in_domain(self):
...
+                )
+            ],
+            item_attributes=[],
+        )
+        self.data_type.save()
+
+        self.data_item = FixtureDataItem(
+            domain=self.domain,
+            data_type_id=self.data_type.get_id,
+            fields= {
```
Comment (dimagimon) [inline corehq/apps/fixtures/tests.py]: E251 unexpected spaces around keyword / parameter equals

File: corehq/apps/fixtures/views.py [hunk 4]
```diff
@@ -8,6 +8,7 @@
 from django.http.response import HttpResponseServerError
 from django.shortcuts import render
 from django.utils.decorators import method_decorator
+from django.utils.safestring import mark_safe
```
Comment (dimagimon) [inline corehq/apps/fixtures/views.py]: F401 'mark_safe' imported but unused

File: corehq/apps/fixtures/views.py [hunk 5]
```diff
@@ -105,6 +107,30 @@ def update_tables(request, domain, data_type_id, test_patch=None):
...
     if request.method == 'POST' or request.method == "PUT":
         fields_update = test_patch or _to_kwargs(request)
+
+        # validate fields
+        validation_errors = []
+        for field_name, options in fields_update['fields'].items():
+            method = options.keys()
+            if 'update' in method:
+                field_name = options['update']
+            if field_name.startswith('xml') and not 'remove' in method:
```
Comment (dimagimon) [inline corehq/apps/fixtures/views.py]: E713 test for membership should be 'not in'

File: corehq/apps/fixtures/views.py [hunk 6]
```diff
@@ -105,6 +107,30 @@ def update_tables(request, domain, data_type_id, test_patch=None):
...
+                validation_errors.append(
+                    _("Field name \"%s\" cannot begin with 'xml'.") % field_name
+                )
+            if is_field_name_invalid(field_name) and not 'remove' in method:
```
Comment (dimagimon) [inline corehq/apps/fixtures/views.py]: E713 test for membership should be 'not in'

**Comment by biyeun** [comment]
@czue updated

**Comment by czue** [comment]
Looks good and thanks for tests! I'm signed off on the XML changes, but I didn't review the new validation stuff closely so will let your buddy or someone else merge.

(In situations like this splitting out changes into separate PRs might make it easier to get them in quicker.)

**Comment by kaapstorm** [comment]
:+1:

**Pull request merged**